Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2024, 2025 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if USE(GBM)
+#include "DMABufBuffer.h"
+#include "GLDisplay.h"
+#include "IntSize.h"
+#include <wtf/OptionSet.h>
+
+struct gbm_bo;
+struct gbm_device;
+typedef void* EGLImage;
+typedef intptr_t EGLAttrib;
+
+namespace WebCore {
+
+class IntPoint;
+class IntRect;
+
+// Use MemoryMappedGPUBuffer to create a OpenGL texture, that's baked by a dma-buf.
+class MemoryMappedGPUBuffer {
+    WTF_MAKE_NONCOPYABLE(MemoryMappedGPUBuffer);
+    WTF_MAKE_FAST_ALLOCATED();
+public:
+    MemoryMappedGPUBuffer(const IntSize&);
+    ~MemoryMappedGPUBuffer();
+
+    enum class BufferFlag : uint8_t {
+        ForceLinear = 1 << 0
+    };
+
+    // Will only return a MemoryMappedGPUBuffer, if gbm_bo allocation + mapping to userland + EGLImage creation succeeded.
+    static std::unique_ptr<MemoryMappedGPUBuffer> create(const IntSize&, OptionSet<BufferFlag>);
+
+    // Map dma-buf into memory, if not yet mapped.
+    bool mapIfNeeded();
+    void unmapIfNeeded();
+
+    // Export gbm_bo buffer as dma-buf and wrap in EGLImage.
+    EGLImage createEGLImageFromDMABuf();
+
+    // Update the underlying data of the dma-buf, as often as desired.
+    // You need to obtain an AccessScope, fencing the write operation.
```

Comment:
Is there a way we can make this constraint explicit? Can we take the scope as an argument, and assert some of its properties -- as we would do if we were passing in a lock?

Reply:
Good idea - better than documentation 😂 will adapt